File: web/lib/review-comparison.ts
/** AI vs human score comparison helpers for evaluator portal. */

import {
  aiScoresToCriteria,
  computeOverallBand,
  CRITERIA_KEYS,
  type HumanCriteriaScores,
} from "@/lib/speaking-band";
import {
  aiScoresToWritingCriteria,
  computeWritingOverallBand,
  WRITING_CRITERIA_KEYS,
  type WritingHumanCriteriaScores,
} from "@/lib/writing-band";

export type CriterionDelta = {
  key: string;
  label: string;
  ai: number | null;
  human: number | null;
  delta: number | null;
  overridden: boolean;
};

export type ScoreComparison = {
  rows: CriterionDelta[];
  aiOverall: number | null;
  humanOverall: number | null;
  deltaOverall: number | null;
  overridden: boolean;
};

const OVERRIDE_THRESHOLD = 0.5;

function roundHalf(value: number): number {
  return Math.round(value * 2) / 2;
}

function deltaBetween(human: number | null, ai: number | null): number | null {
  if (human == null || ai == null) return null;
  return roundHalf(human - ai);
}

export function compareSpeakingScores(
  human: Partial<HumanCriteriaScores> | null | undefined,
  aiScores: Record<string, unknown> | null | undefined,
  labels: Record<keyof HumanCriteriaScores, string>,
): ScoreComparison {
  const ai = aiScoresToCriteria(aiScores);
  const rows: CriterionDelta[] = CRITERIA_KEYS.map((key) => {
    const aiVal = ai?.[key] ?? null;
    const humanVal =
      human && human[key] != null ? Number(human[key]) : null;
    const delta = deltaBetween(humanVal, aiVal);
    return {
      key,
      label: labels[key],
      ai: aiVal,
      human: humanVal,
      delta,
      overridden: delta != null && Math.abs(delta) >= OVERRIDE_THRESHOLD,
    };
  });
  const aiOverall = ai ? computeOverallBand(ai) : null;
  const humanOverall =
    human && CRITERIA_KEYS.every((k) => human[k] != null)
      ? computeOverallBand(human as HumanCriteriaScores)
      : null;
  const deltaOverall = deltaBetween(humanOverall, aiOverall);
  return {
    rows,
    aiOverall,
    humanOverall,
    deltaOverall,
    overridden:
      rows.some((r) => r.overridden) ||
      (deltaOverall != null && Math.abs(deltaOverall) >= OVERRIDE_THRESHOLD),
  };
}

export function compareWritingScores(
  human: Partial<WritingHumanCriteriaScores> | null | undefined,
  aiScores: Record<string, unknown> | null | undefined,
  labels: Record<keyof WritingHumanCriteriaScores, string>,
): ScoreComparison {
  const ai = aiScoresToWritingCriteria(aiScores);
  const rows: CriterionDelta[] = WRITING_CRITERIA_KEYS.map((key) => {
    const aiVal = ai?.[key] ?? null;
    const humanVal =
      human && human[key] != null ? Number(human[key]) : null;
    const delta = deltaBetween(humanVal, aiVal);
    return {
      key,
      label: labels[key],
      ai: aiVal,
      human: humanVal,
      delta,
      overridden: delta != null && Math.abs(delta) >= OVERRIDE_THRESHOLD,
    };
  });
  const aiOverall = ai ? computeWritingOverallBand(ai) : null;
  const humanOverall =
    human && WRITING_CRITERIA_KEYS.every((k) => human[k] != null)
      ? computeWritingOverallBand(human as WritingHumanCriteriaScores)
      : null;
  const deltaOverall = deltaBetween(humanOverall, aiOverall);
  return {
    rows,
    aiOverall,
    humanOverall,
    deltaOverall,
    overridden:
      rows.some((r) => r.overridden) ||
      (deltaOverall != null && Math.abs(deltaOverall) >= OVERRIDE_THRESHOLD),
  };
}

export function formatDelta(delta: number | null): string {
  if (delta == null) return "—";
  if (delta === 0) return "0";
  return delta > 0 ? `+${delta.toFixed(1)}` : delta.toFixed(1);
}

File: api/review_comparison.py
"""AI vs human score comparison helpers for approve audit + analytics."""

from __future__ import annotations

from typing import Any

OVERRIDE_THRESHOLD = 0.5


def round_half(value: float) -> float:
    return round(value * 2) / 2


def delta_between(human: float | None, ai: float | None) -> float | None:
    if human is None or ai is None:
        return None
    return round_half(human - ai)


def is_overridden(
    *,
    human_criteria: dict[str, Any] | None,
    ai_criteria: dict[str, Any] | None,
    human_overall: float | None = None,
    ai_overall: float | None = None,
) -> bool:
    """True if any criterion |Δ| ≥ 0.5 or overall |Δ| ≥ 0.5."""
    if human_criteria and ai_criteria:
        for key, human_val in human_criteria.items():
            if human_val is None or key not in ai_criteria or ai_criteria[key] is None:
                continue
            d = delta_between(float(human_val), float(ai_criteria[key]))
            if d is not None and abs(d) >= OVERRIDE_THRESHOLD:
                return True
    d_overall = delta_between(human_overall, ai_overall)
    return d_overall is not None and abs(d_overall) >= OVERRIDE_THRESHOLD


def approve_audit_metadata(
    *,
    human_band: float,
    human_criteria: dict[str, Any],
    ai_band: float | None,
    ai_criteria: dict[str, Any] | None,
    extra: dict[str, Any] | None = None,
) -> dict[str, Any]:
    delta_overall = delta_between(human_band, ai_band)
    meta: dict[str, Any] = {
        "human_band": human_band,
        "human_criteria_scores": human_criteria,
        "ai_band": ai_band,
        "ai_criteria": ai_criteria,
        "overridden": is_overridden(
            human_criteria=human_criteria,
            ai_criteria=ai_criteria,
            human_overall=human_band,
            ai_overall=ai_band,
        ),
        "delta_overall": delta_overall,
    }
    if extra:
        meta.update(extra)
    return meta


def mae(values: list[float]) -> float | None:
    if not values:
        return None
    return round(sum(abs(v) for v in values) / len(values), 3)
